_posts/2020-11-26-torch-brain-segmentation/torch_brain_segmentation.Rmd

```diff
@@ -0,0 +1,705 @@
+
+![](images/unet.png "The original U-Net, as depicted in Ronnerberger et al. (2015).")
+
+In a nutshell, the left-hand side of the U resembles the convolutional architectures used in image classification: It successively reduces spatial resolution. At the same time, another dimension -- the *channels* dimension -- is used to build up a hierarchy of features, ranging from very basic to very specialized.
+
+Unlike in classification, however, the output should have the same spatial resolution as the input. Thus, we need to upsize again -- this is taken care of by the right-hand side of the U. But: How are we going to arrive at a good *per-pixel* classification, now that so much spatial information has been lost?
+
+This is what the "bridges" are for: At each level, the input to an upsampling layer is a *concatenation* of the previous layer's output -- which went through the whole compression/decompression routine -- and some preserved intermediate representation from the downsizing phase. In this way, a U-Net architecture combines attention to detail with feature extraction.
+
+## Brain image segmentation
+
+With U-Net, domain applicability is as broad as the architecture is flexible. Here, we want to detect abnormalities in brain scans. The dataset, used in @BUDA2019218, contains MR images together with manually created [FLAIR](https://en.wikipedia.org/wiki/Fluid-attenuated_inversion_recovery) abnormality segmentation masks. It is available on [Kaggle](https://www.kaggle.com/mateuszbuda/lgg-mri-segmentation).
+
+Nicely, the paper is accompanied by a [GitHub repository](https://github.com/mateuszbuda/brain-segmentation-pytorch). Below, we closely follow (though not exactly replicate) the authors' preprocessing and data augmentation code.
+
+As is often the case in medical imaging, there is notable class imbalance in the data. For every patient, sections have been taken at multiple positions. (Number of sections per patient varies.) Most sections do not exhibit any lesions; the corresponding masks are colored black everywhere.
+
+Here are three examples where the masks *do* indicate abnormalities:
+
+![](images/scans.png)
+
+Let's see if we can build a U-Net that generates such masks for us.
+
+## Data
+
+Before you start typing, here is a [Colaboratory notebook](https://colab.research.google.com/drive/1gcPOlF4jff42yiquRbbCxnRANEqPpGMD?usp=sharing) to conveniently follow along.
+
+We use `pins` to obtain the data. Please see [this introduction](https://pins.rstudio.com/articles/boards-kaggle.html) if you haven't been using that package before.
+
+```{r}
+library(torch)
+library(torchvision)
+library(tidyverse)
+library(magick)
+library(zeallot)
+library(cowplot)
+library(zip)
+library(pins)
```

Comment:
Totally your choice, but when there are this many libraries, I find it helpful when the authors comment what each is being used for (especially when the names are non-obvious). If it's hard to sum up, no big deal—just a thought.

Reply:
makes sense!